### ISSUE TITLE
thermal sensor support for ZTEX 1.15y "Rev 2"

### ISSUE THREAD

**Issue by roycewilliams** [magnumripper/JohnTheRipper#2338]
This is a feature request -- a "nice-to-have" for the recent ZTEX 1.15y work by @Apingis.

Onboard thermal sensors were added in the "Rev 2" version of the boards. These boards appear to be common, and it appears that many of the licensed clone boards are also of the "Rev 2" variety.

This is a request to tie these sensors into the existing thermal framework.

Here's a thread where ZTEX themselves announces a newer version of their miner that takes advantage of the new sensors:

https://bitcointalk.org/index.php?topic=40047.msg1338948#msg1338948

**Comment by roycewilliams** [comment]
The ZTEX FX2 framework includes ztex/fx2/ztex-temp1.h. 

Other than the copyright date, this forked file is identical to the most recent one available at ztex.de.

https://github.com/vpereira/golden_unicorn/blob/master/ztex/include/ztex-temp1.h

**Comment by Apingis** [comment]
Board firmware already has support for temp. sensor, that's enabled by default. (Actual presence of the sensor can be checked with FWLoader -ii). 
Boards in my posession don't have sensors, well it wouldn't be hard to add such support, just will need somebody to test.

The issue is what would JtR do if some board overheats. There's no control over power supply. It can stop using the board for processing to reduce power consumption. Additionally it can remove bitstreams, without bitstreams heatsinks are cold.

Questions:
1. Where can I see "the existing thermal framework"?
2. There probably must be a hook invoked when a user hits 'any key' so device-specific format outputs device-specific data? There can be data other than SN and temperature, such as I can add fpga utilization.

**Comment by magnumripper** [comment]
The "existing framework" is in common-gpu.c. It applies to CUDA and/or OpenCL right now and only supports AMD and nvidia (and currently only Linux simply because I don't touch Windows ever - should be easy to fix for someone else). The current code for GPU overheat protection simply aborts john, the temperature limit is in john.conf.

There's code in status.c that can show (according to john.conf settings) temp, fan speed and/or utilisation upon keypress. That too is a bit of a hack right now.

**Comment by roycewilliams** [comment]
Per-chip granularity might be desirable. There are four chips, each with its own sensor (and often its own  fan). Failure of one fan can result in a temperature spike for a single chip, without affecting the other chips. If disabling processing on individual chips is easy, that would be nice.  If it's not easy, that could be deferred. :)

Either way, I have a board that we can use to test whatever support is added.

**Comment by magnumripper** [comment]
The current code has per-GPU granularity. We could alter it (so we still have a single standard interface) or you could simply return the highest among the 4.

**Comment by roycewilliams** [comment]
Being able to tell which chip(s) are running hot would be operationally useful.

@Apingis, I know that you were busy the past couple of months; how are things looking now?

**Comment by roycewilliams** [comment]
In the short term, just displaying the current temperature would be a big help, operationally speaking. Today, the only way that I can troubleshoot thermal issues is to stop john, and then start up a bitcoin miner that is aware of the sensors and can show the temperatures.

**Comment by solardiz** [comment]
Royce, are you actually running into thermal issues when running descrypt-ztex?

**Comment by roycewilliams** [comment]
Yes, but it is not john's fault - it's my process of troubleshooting physical cooling issues.
